src/data/repositories/ProgramLoggerD2Repository.ts

```diff
@@ -0,0 +1,172 @@
+import { D2Api, MetadataPick, D2TrackerEvent, DataValue } from "../../types/d2-api";
+import { apiToFuture, FutureData } from "../api-futures";
+import { Future } from "../../domain/entities/generic/Future";
+import { Id } from "../../domain/entities/Base";
+import { Log } from "../../domain/entities/Log";
+import { ProgramLoggerConfig } from "../../domain/entities/LoggerConfig";
+import { LoggerRepository } from "../../domain/repositories/LoggerRepository";
+
+const GLOBAL_ORGANISATION_UNIT_ID = "H8RixfF8ugH";
+const IMPORT_STRATEGY_CREATE = "CREATE";
+const TRACKER_IMPORT_JOB = "TRACKER_IMPORT_JOB";
+const EVENT_PROGRAM_STATUS = "COMPLETED";
+
+export class ProgramLoggerD2Repository implements LoggerRepository {
+    api: D2Api;
+    organisationUnitId: Id;
+    programId: Id;
+    messageId: Id;
+    messageTypeId: Id;
+
+    constructor(config: ProgramLoggerConfig) {
+        const {
+            baseUrl,
+            auth,
+            programId,
+            loggerDataElements,
+            organisationUnitId = GLOBAL_ORGANISATION_UNIT_ID,
+        } = config;
+
+        this.api = auth
+            ? new D2Api({ baseUrl: baseUrl, auth: auth })
+            : new D2Api({ baseUrl: baseUrl });
```

Comment:
This check should not be needed, if auth is undefined, it will be send as undefined -> `new D2Api({ baseUrl: baseUrl, auth: auth })`